File: src/main/scala/Day07/Day07.hs
import Data.List.Split
import Data.List

main = do
  ls <- readFile "Day07/input"
  let array::[Int] = sort $ map (read) $ splitOn "," ls
  print $ part1 array
  print $ part2 array

part1::[Int] -> Int
part1 crabs = 
  sum $ map (\crab -> abs(crab-median)) crabs
  where 
    median = crabs!!(length crabs `div`2)

part2::[Int] -> Int
part2 crabs = 
  (minMove crabs `div` 2)

move crabs pos = sum $ map (\x -> let a = abs (x - pos) in a * (a + 1)) crabs

minMove crabs = minimum (map (move crabs) [0..1000])    

File: src/main/scala/Day07/Day07.scala
package Day07

import scala.io.Source

object Day07 {

    def main(args: Array[String]): Unit = {
        val source = Source.fromFile("src/main/scala/Day07/input")
        val input = source.mkString.trim.split(",").map(_.toInt).toVector.sorted

        println(part1(input) + " (correct: 356958)")
        println(part2(input) + " (correct: 105461913)")
    }

    def part1(crabs: Vector[Int]): Int = {
        val med = crabs(crabs.size / 2)
        crabs.map(crab => (crab - med).abs).sum
    }

    def part2(crabs: Vector[Int]): Long = {
        (crabs.min to crabs.max).map(
            pos => crabs.map(
                crab => summe((pos - crab).abs)
            ).sum
        ).min
    }

    def summe(n: Int): Int = n * (n + 1) / 2

}